### DISCUSSION TITLE
set_committed_value(obj,'xyz',...) vs directly setting scalar / set_attribute()

### DISCUSSION THREAD

**Discussion by aathan** [sqlalchemy/sqlalchemy#8238] [Usage Questions]
I surmise from the documentation of methods such as is_modified() and set_attribute() that instrumented attributes maintain some kind of value history that is relevant to heuristics that can reduce emitted SQL in certain instances; and that using set_committed_value(obj,'xyz',value) vs obj.xyz=value clears history and sets the value "as it is in the DB."

Do I have that right? Are there any cases where obj.xyz=value may be the better choice?

I have an object that contains a messages[] collection as a one to many relationship. Those messages are specialized and the object hangs some of those onto scalar attributes for quicker access.  E.g.

obj.messages = list( object_instance_of_type_A, object_instance_of_type_B, etc...)
obj.message_A = object_instance_of_type_A

If obj.message_A is declared as a relationship as well as obj.messages, then I imagine access to the scalar attribute message_A would involve a separate fetch even though message_A is in the messages collection. So, my approach is to declare as lazy=noload the object_A relationship (vs perhaps not defining a relationship on that attribute at all?)

Then, in an @event.load(...) handler for obj, I traverse obj.messages and individually set the individual attributes (and the back pointer):

```
@event.load(Obj,'load')
def fixup_load(obj,...):
   for m in obj.messages:
      if ...:
          set_committed_value(m,'owner',obj)
          obj.object_A = m
```

Should I not define any relationships for the object_A's (instead of defining them as noloads)?
If not, then I guess obj.object_A = m above is fine. However, if the object_A attribute is known to SQLA (is an instrumented attribute), should I instead be writing set_committed_value(obj,'object_A',m) above?

Thanks!

**Comment by zzzeek** [comment]
> I surmise from the documentation of methods such as is_modified() and set_attribute() that instrumented attributes maintain some kind of value history that is relevant to heuristics that can reduce emitted SQL in certain instances; and that using set_committed_value(obj,'xyz',value) vs obj.xyz=value clears history and sets the value "as it is in the DB."
> 
> Do I have that right? Are there any cases where obj.xyz=value may be the better choice?

set_committed_value() is only used when implementing custom caching routines to pull objects from a cache, or when implementing custom attribute load schemes (which seems to be what you're doing).  otherwise it is not appropriate to use this function.

> 
> I have an object that contains a messages[] collection as a one to many relationship. Those messages are specialized and the object hangs some of those onto scalar attributes for quicker access. E.g.
> 
> obj.messages = list( object_instance_of_type_A, object_instance_of_type_B, etc...) obj.message_A = object_instance_of_type_A
> 
> If obj.message_A is declared as a relationship as well as obj.messages, then I imagine access to the scalar attribute message_A would involve a separate fetch even though message_A is in the messages collection. So, my approach is to declare as lazy=noload the object_A relationship (vs perhaps not defining a relationship on that attribute at all?)

you can use lazy=noload but keep in mind if you have code that accesses this attribute and you have not placed a value there, it will return an empty collection which means your application is getting the wrong answer for a particular attribute.   people usually use this because they are stuck using some kind of serialization utility where they don't care about the value of some relationship yet they still want to have it there (again, it's already a program where they have a lot of red flags happening IMO).    if you dont want to lazy load things ever, lazy="raise_on_sql" or "raise" is safer from a data integrity point of view because your application will never proceed with invalid data.


> 
> Then, in an @event.load(...) handler for obj, I traverse obj.messages and individually set the individual attributes (and the back pointer):
> 
> ```
> @event.load(Obj,'load')
> def fixup_load(obj,...):
>    for m in obj.messages:
>       if ...:
>           set_committed_value(m,'owner',obj)
>           obj.object_A = m
> ```

if you're doing a custom load scheme and "obj" is also a mapped object, I'd use set_committed_value there also most likely.

> 
> Should I not define any relationships for the object_A's (instead of defining them as noloads)? 

if you truly have a whole setup to stick collections onto things that you are loading elsewhere and you have no mutability in play, then yes, because it seems like your application is really not an ORM app at all and if you just had a big bunch of Core result objects that you placed into simple non-mapped objects as big read-only, in memory database, you'd be much better off.   Core has no mutable state or any of the things that are causing you problems.

> If not, then I guess obj.object_A = m above is fine. However, if the object_A attribute is known to SQLA (is an instrumented attribute), should I instead be writing set_committed_value(obj,'object_A',m) above?

yes use set_committed_value() if you are building up a fixed view of database data that you are loading separately.

**Reply by aathan** [reply]
Regarding "core vs orm" I'm still consuming the docs in more detail but I'm currently under the impression (er, assumption) that generating SQL for loading related objects is super easy in ORM and not so easy in Core. Maybe the Query SQL generator can easily be used vs Core instead of via Session?

In order to avail myself nice ORM query generation features my first pass is to use ORM so I can emit `option(joinedload('*'))` queries; then, in an @event.load() I do any scalar "pointer" fixups out of the loaded collections etc per above.

I'm aware of `lazy=raise` and `lazy=raise_on_sql` but for the moment not raising is more important than having complete (typically historical) data in some collections. I am using raise in dev/staging only, with the goal of getting to the point where `lazy=noload` is in fact equivalent (because no loads are happening). `lazy=raise` would require implementing operation restart after a raise from "anywhere", which is non-trivial given the breadth of the app. Any resulting data errors can be healed ex-post-facto.

To your point, I'm currently chasing down one spot where the `obj.messages` collection is ending up empty even though other parts of the app have most certainly put objects into it. Frustratingly, I haven't been able to reproduce the behavior in dev. Either a new proxy with empty collection has been loaded and all the logs make it look like the original (because they log `obj.id_column` but not hex(id(obj)), or the collection has been reset to empty by the ORM perhaps on a session.close() or expunge()? Strangely, the contents of the collection make it to the DB so if the ORM is resetting the collection to empty on the existing proxy object, I'm scratching my head as to why.

**Reply by zzzeek** [reply]
> Regarding "core vs orm" I'm still consuming the docs in more detail but I'm currently under the impression (er, assumption) that generating SQL for loading related objects is super easy in ORM and not so easy in Core. Maybe the Query SQL generator can easily be used vs Core instead of via Session?

sure, that's one of the things the ORM offers is the relationship SQL generation.     This works most easily when you have mapped classes with relationships on them, sure.    There's not much way to use them in Core without having mapped classes already.


> 
> In order to avail myself nice ORM query generation features my first pass is to use ORM so I can emit `option(joinedload('*'))` queries; then, in an @event.load() I do any scalar "pointer" fixups out of the loaded collections etc per above.

yes you'd have to do something like that.    

> 
> I'm aware of `lazy=raise` and `lazy=raise_on_sql` but for the moment not raising is more important than having complete (typically historical) data in some collections.

yes if you are fine with empty collections that are wrong, lazy='noload' is fine.

> I am using raise in dev/staging only, with the goal of getting to the point where `lazy=noload` is in fact equivalent (because no loads are happening). `lazy=raise` would require implementing operation restart after a raise from "anywhere", which is non-trivial given the breadth of the app. Any resulting data errors can be healed ex-post-facto.

if you are truly setting up collections everywhere there was a lazy=noload, I would think you could switch to lazy='raise` and that would basically be an assertion layer.  you wouldn't have any raises if you aren't accessing the collections unloaded.
